File: packages/hydra-kernel/src/main/haskell/Hydra/Sources/Kernel/Lib/Effects.hs
-- | Primitive declarations for the hydra.lib.effects namespace.

module Hydra.Sources.Kernel.Lib.Effects where

import Hydra.Kernel
import qualified Hydra.Overlay.Haskell.Bootstrap     as Bootstrap
import           Hydra.Overlay.Haskell.Dsl.Typed.Phantoms as Phantoms hiding (apply, compose, map)
import qualified Hydra.Overlay.Haskell.Dsl.Types         as Types
import           Hydra.Overlay.Haskell.Dsl.Types         (effect)
import           Hydra.Sources.Kernel.Types.All
import           Prelude hiding ((++), foldl, map, pure)


ns :: ModuleName
ns = ModuleName "hydra.lib.effects"

module_ :: Module
module_ = Module {
            moduleName = ns,
            moduleDefinitions = DefinitionPrimitive <$> definitions,
            moduleDependencies = Bootstrap.unqualifiedDep <$> kernelTypesModuleNames,
            moduleMetadata = Bootstrap.descriptionMetadata (Just "Primitives in the hydra.lib.effects module.")}
  where
    definitions = [apply, bind, compose, foldList, map, mapList, mapOptional, mapSet, pure]

define :: String -> String -> TermSignature -> [String] -> PrimitiveDefinition
define = impurePrimitiveInModule module_

-- Shared type variables
tx, ty, tz :: Type
tx = Types.var "x"
ty = Types.var "y"
tz = Types.var "z"

apply :: PrimitiveDefinition
apply = define "apply" "Applicative apply for effects."
  (sigWithParams [("ef", "the effect producing the function to apply"),
                  ("ex", "the effect producing the argument")]
    $ TypeScheme [Name "x", Name "y"]
    (effect (tx Types.~> ty) Types.~> effect tx Types.~> effect ty) Nothing)
  ["apply(ef, ex) describes an effectful computation which first interprets ef to produce a\
  \ function, then interprets ex to produce an argument, and returns the result of applying the\
  \ function to the argument."]

bind :: PrimitiveDefinition
bind = define "bind" "Sequence two effectful computations."
  (sigWithParams [("e", "the effect to interpret first"),
                  ("f", "the function producing the next effect from e's result")]
    $ TypeScheme [Name "x", Name "y"]
    (effect tx Types.~> (tx Types.~> effect ty) Types.~> effect ty) Nothing)
  ["bind(e, f) describes an effectful computation which first interprets e, then passes its\
  \ result to f to obtain the next effect.",
   "Host interpreters provide the operational meaning; Hydra uses the type effect<t> to keep the\
   \ effect boundary explicit."]

compose :: PrimitiveDefinition
compose = define "compose" "Kleisli composition for effects."
  (sigWithParams [("f", "the first effect-returning function"),
                  ("g", "the second effect-returning function"),
                  ("x", "the input to the composed computation")]
    $ TypeScheme [Name "x", Name "y", Name "z"]
    ((tx Types.~> effect ty) Types.~> (ty Types.~> effect tz) Types.~> tx Types.~> effect tz) Nothing)
  ["compose(f, g, x) describes an effectful computation equivalent to bind(f(x), g)."]

foldList :: PrimitiveDefinition
foldList = define "foldList" "Left-fold over a list with an effect-returning function."
  (sigWithParams [("f", "the effect-returning combining function"),
                  ("acc0", "the initial accumulator value"),
                  ("xs", "the list to fold over")]
    $ TypeScheme [Name "x", Name "y"]
    ((tx Types.~> ty Types.~> effect tx) Types.~> tx Types.~> Types.list ty Types.~> effect tx) Nothing)
  ["foldList(f, acc0, xs) describes an effectful left fold over xs, sequencing applications of f from\
  \ left to right."]

map :: PrimitiveDefinition
map = define "map" "Map a pure function over the result of an effect."
  (sigWithParams [("f", "the pure function to apply to the effect's result"),
                  ("e", "the effect to interpret")]
    $ TypeScheme [Name "x", Name "y"]
    ((tx Types.~> ty) Types.~> effect tx Types.~> effect ty) Nothing)
  ["map(f, e) describes an effectful computation which interprets e and applies f to its result.",
   "Equivalent to bind(e, \\x -> pure(f x))."]

mapList :: PrimitiveDefinition
mapList = define "mapList" "Map an effect-returning function over a list."
  (sigWithParams [("f", "the effect-returning function to apply to each element"),
                  ("xs", "the list to map over")]
    $ TypeScheme [Name "x", Name "y"]
    ((tx Types.~> effect ty) Types.~> Types.list tx Types.~> effect (Types.list ty)) Nothing)
  ["mapList(f, xs) describes an effectful computation which applies f to each element of xs from\
  \ left to right and collects the results."]

mapOptional :: PrimitiveDefinition
mapOptional = define "mapOptional" "Map an effect-returning function over an optional."
  (sigWithParams [("f", "the effect-returning function to apply to the contained value"),
                  ("m", "the optional to map over")]
    $ TypeScheme [Name "x", Name "y"]
    ((tx Types.~> effect ty) Types.~> Types.optional tx Types.~> effect (Types.optional ty)) Nothing)
  ["mapOptional(f, m) describes an effectful computation which returns none when m is none, or\
  \ applies f to the contained value and wraps the result in given."]

mapSet :: PrimitiveDefinition
mapSet = define "mapSet" "Map an effect-returning function over a set."
  (sigWithParams [("f", "the effect-returning function to apply to each element"),
                  ("s", "the set to map over")]
    $ Types.polyConstrained [("x", [Name "ordering"]), ("y", [Name "ordering"])]
    ((tx Types.~> effect ty) Types.~> Types.set tx Types.~> effect (Types.set ty)))
  ["mapSet(f, s) describes an effectful computation which applies f to each element of s, sequencing\
  \ the effects in ascending element order (the order is normative, since the effects are observable),\
  \ and collects the results as a set, deduplicated by the result type's ordering.",
   "Requires 'ordering' constraints on both element types (the set type contract)."]

pure :: PrimitiveDefinition
pure = define "pure" "Lift a pure value into an effect."
  (sigWithParams [("x", "the pure value to lift into an effect")]
    $ TypeScheme [Name "x"] (tx Types.~> effect tx) Nothing)
  ["pure(x) describes an effectful computation which succeeds with x without requiring host\
  \ interaction."]

File: overlay/clojure/hydra-kernel/src/main/clojure/hydra/overlay/clojure/lib/effects.clj
(ns hydra.overlay.clojure.lib.effects)

;; Clojure implementations of hydra.lib.effects primitives (#494).
;;
;; The Hydra type effect<t> is transparent in the Lisp dialects (effect<t> = t): there is no
;; TypeVariantEffect in the target, so effectful programs are ordinary eager native code and
;; "running the effect" simply means forcing the value. These primitives therefore reduce to
;; ordinary applications. See the Haskell reference implementation in Hydra.Haskell.Lib.Effects
;; (where effect<t> = IO t). All functions are curried, matching the Clojure prim runtime style.
;;
;; Optional representation (term-level): (list :none) or (list :given val).

;; apply :: effect<x -> y> -> effect<x> -> effect<y>
(def hydra_lib_effects_apply
  "Applicative apply for effects. Since effects are transparent, this just applies ef to ex."
  (fn [ef] (fn [ex] (ef ex))))

;; bind :: effect<x> -> (x -> effect<y>) -> effect<y>
(def hydra_lib_effects_bind
  "Sequence two effectful computations. Since effects are transparent, this just applies f to a."
  (fn [a] (fn [f] (f a))))

;; compose :: (x -> effect<y>) -> (y -> effect<z>) -> x -> effect<z>
(def hydra_lib_effects_compose
  "Kleisli composition for effects: run f, then g on its result."
  (fn [f] (fn [g] (fn [x] (g (f x))))))

;; fold_list :: (x -> y -> effect<x>) -> x -> [y] -> effect<x>
(def hydra_lib_effects_fold_list
  "Left-fold over a list with an effect-returning function."
  (fn [f] (fn [acc] (fn [xs]
    (reduce (fn [a x] ((f a) x)) acc xs)))))

;; map :: (x -> y) -> effect<x> -> effect<y>
(def hydra_lib_effects_map
  "Map a pure function over the result of an effect. Since effects are transparent, just apply f."
  (fn [f] (fn [a] (f a))))

;; mapList :: (x -> effect<y>) -> [x] -> effect<[y]>
(def hydra_lib_effects_map_list
  "Map an effect-returning function over a list, collecting the results."
  (fn [f] (fn [xs] (mapv f xs))))

;; mapOptional :: (x -> effect<y>) -> Maybe x -> effect<Maybe y>
(def hydra_lib_effects_map_optional
  "Map an effect-returning function over an optional."
  (fn [f] (fn [m]
    (cond
      (or (nil? m)
          (and (sequential? m) (empty? m))
          (and (sequential? m) (= (first m) :none)))
      (list :none)

      (and (sequential? m) (= (first m) :given))
      (list :given (f (second m)))

      ;; bare value (term-level maybe)
      :else
      (list :given (f m))))))

;; pure :: x -> effect<x>
(def hydra_lib_effects_pure
  "Lift a pure value into an effect. Since effects are transparent, this is the identity."
  (fn [x] x))
